File: apps/financas/scripts/insight.mjs
#!/usr/bin/env node
//
// Insight de IA -> le os numeros ja calculados (GET /api/insights/numbers),
// manda pro Ollama local escrever uma leitura em texto por cima deles, e
// publica o resultado (POST /api/insights).
//
// Spec: docs/superpowers/specs/2026-07-28-financas-ui-insights-design.md §3
// Backend (Task 3): apps/financas/src/domain/insights.ts + routes/insights.ts
//
// O Mac EMPURRA, o app LE (spec §3) — nenhuma tela chama este comando, e
// nenhum servidor precisa estar de pe alem do Ollama rodando NESTA maquina
// durante a execucao. O dono roda isto quando quer uma leitura fresca do
// mes; a tela mostra o que foi publicado por ultimo, com a data.
//
// Zero custo de AI: Ollama local, nunca Workers AI (medido funcionando
// nesta conta e DESCARTADO por ser pago — ver §0 do spec).
//
// ⚠️ O MODELO NUNCA VE LANCAMENTO CRU. `GET /api/insights/numbers` devolve
// só agregado (totais por categoria/periodo, ja calculados no Worker via
// `byCategory`) — o prompt (buildPrompt, abaixo) so recebe esses agregados,
// nunca uma linha do livro-caixa. Isso mantem o escopo do INGEST_TOKEN
// honesto mesmo apos a extensao desta task (le totais, escreve prosa,
// nunca toca o livro-caixa — ver CLAUDE.md § Insight de IA).
//
// ⚠️ Temperatura zero (`options.temperature: 0` em callOllama) — isto e
// extracao/resumo de fatos ja calculados, nao criacao; variacao entre
// execucoes aqui e defeito, mesma regra de scripts/pdf-import.mjs.
//
// Disciplina de erro (mesmo padrao de pdf-import.mjs e backup-d1.sh —
// "a mensagem de erro e o produto"):
//   - Ollama desligado -> como iniciar, nunca ECONNREFUSED cru
//   - modelo nao instalado -> o comando exato de "ollama pull"
//   - INGEST_TOKEN ausente do ambiente -> diz isso, e como configurar
//   - API inalcancavel (rede) != API recusou (401/4xx/5xx) — mensagens
//     diferentes pra causas diferentes
//   - modelo devolve texto vazio -> falha alto, com a saida bruta no log;
//     nunca publica insight vazio silenciosamente

import { pathToFileURL } from 'node:url'
import { formatBRL } from '@piluvitu/tools/money'

export const DEFAULT_MODEL = 'qwen2.5:7b-instruct'
export const DEFAULT_OLLAMA_URL = 'http://localhost:11434/api/generate'
// Mesmo host que serve a SPA e a API (ver CLAUDE.md § Deploy) — UI e API
// vivem no mesmo domínio, sem NEXT_PUBLIC_API_URL nem base configuravel.
export const DEFAULT_API_URL = 'https://financas.piluvitu.com.br'

const COMPETENCE_RE = /^\d{4}-(0[1-9]|1[0-2])$/

// Teresina e UTC-3 FIXO (o Piaui nao adota horario de verao desde 2019) —
// mesma constante/formula de src/lib/dates.ts (Worker) e web/src/lib/dates.ts
// (SPA), duplicada aqui de proposito: este CLI e um TERCEIRO runtime (Node
// puro, rodando no Mac do dono), sem fronteira de import pros outros dois
// (mesmo raciocinio ja documentado nos outros dois arquivos).
const TERESINA_OFFSET_MS = 3 * 60 * 60 * 1000

export function competenciaAtual(now = new Date()) {
  return new Date(now.getTime() - TERESINA_OFFSET_MS).toISOString().slice(0, 7)
}

// -------------------------------------------------------------------------
// 1. Prompt — só recebe os agregados de `numbers` (InsightNumbers, ver
//    domain/insights.ts), nunca um lançamento cru. Formata dinheiro sempre
//    via formatBRL (@piluvitu/tools/money) — nunca float solto no texto.
// -------------------------------------------------------------------------

function formatCategoryLine(row, index) {
  return `${index + 1}. ${row.category_name}: ${formatBRL(Math.abs(row.total_cents))}`
}

function linhaVariacao(numbers) {
  const { variation_cents, variation_pct, previous_competence } = numbers
  if (variation_pct === null) {
    return `Sem base de comparação: não houve gasto registrado em ${previous_competence}.`
  }
  if (variation_cents === 0) {
    return `Sem variação em relação a ${previous_competence}.`
  }
  const direcao = variation_cents > 0 ? 'Aumento' : 'Redução'
  return `${direcao} de ${formatBRL(Math.abs(variation_cents))} (${Math.abs(variation_pct)}%) em relação a ${previous_competence}.`
}

function linhaMaiorCrescimento(numbers) {
  const b = numbers.biggest_increase
  if (b === null) {
    return '(sem dado suficiente para apontar)'
  }
  const direcao = b.delta_cents >= 0 ? 'aumento' : 'redução'
  return (
    `${b.category_name}: foi de ${formatBRL(Math.abs(b.previous_cents))} ` +
    `para ${formatBRL(Math.abs(b.current_cents))} (${direcao} de ` +
    `${formatBRL(Math.abs(b.delta_cents))}).`
  )
}

/**
 * Monta o prompt a partir de `InsightNumbers` (o payload de
 * `GET /api/insights/numbers`) — nada além disso entra aqui. As "REGRAS
 * OBRIGATÓRIAS" no fim são o que impede o modelo de inventar um número
 * (spec §3: "nenhum número exibido pode vir do modelo" — a tela, Task 5,
 * renderiza os números a partir do dado; o texto que este prompt produz é
 * só a leitura ao redor).
 */
export function buildPrompt(numbers) {
  const {
    competence,
    previous_competence,
    top_categories,
    total_cents,
    previous_total_cents,
  } = numbers

  const linhasCategorias =
    top_categories.length > 0
      ? top_categories.map(formatCategoryLine).join('\n')
      : '(nenhum gasto registrado nesta competência)'

  return `Você escreve um resumo financeiro curto, em português do Brasil, para o dono de um controle financeiro pessoal/PJ.

Abaixo estão os ÚNICOS fatos que você pode usar. Eles já foram calculados por consulta exata ao banco de dados — você NÃO tem acesso a nenhum lançamento individual, só a estes totais já agregados.

Competência: ${competence} (mês anterior de comparação: ${previous_competence})
Total gasto em ${competence}: ${formatBRL(Math.abs(total_cents))}
Total gasto em ${previous_competence}: ${formatBRL(Math.abs(previous_total_cents))}
Variação: ${linhaVariacao(numbers)}

Maiores categorias de gasto em ${competence} (da maior para a menor):
${linhasCategorias}

Categoria com a maior variação de gasto em relação a ${previous_competence}:
${linhaMaiorCrescimento(numbers)}

Escreva um parágrafo de 3 a 5 frases resumindo esses fatos para o dono, em tom direto, sem saudação, sem "espero que ajude", sem markdown, sem lista — só o parágrafo corrido.

REGRAS OBRIGATÓRIAS, sem exceção:
- Use SOMENTE os números e nomes de categoria que aparecem acima. Não calcule nada novo, não arredonde diferente do que já está escrito, não converta nem estime.
- NUNCA invente um valor, uma categoria ou uma porcentagem que não esteja listada acima.
- Se um fato não tem dado suficiente (ex.: "sem base de comparação" ou "nenhum gasto registrado"), diga isso em vez de inventar um número.
- Não é preciso citar todos os números — escolha o que mais importa para contar a história do mês, mas nunca cite um número que não veio da lista acima.`
}

// -------------------------------------------------------------------------
// 2. Ollama — mesmo padrão de erro de scripts/pdf-import.mjs (duplicado de
//    propósito: são dois CLIs independentes, sem um módulo compartilhado
//    entre eles hoje — mesma decisão já tomada pros mirrors de datas do
//    Worker/SPA).
// -------------------------------------------------------------------------

function erroDeConexaoRecusada(err) {
  return err?.cause?.code === 'ECONNREFUSED' || err?.code === 'ECONNREFUSED'
}

export async function callOllama({ model, prompt, url, fetchImpl }) {
  let response
  try {
    response = await fetchImpl(url, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({
        model,
        prompt,
        stream: false,
        // Temperatura zero: extração/resumo de fatos já calculados não é
        // criação — variação entre execuções aqui é defeito.
        options: { temperature: 0 },
      }),
    })
  } catch (err) {
    if (erroDeConexaoRecusada(err)) {
      throw new Error(
        `não consegui conectar ao Ollama em ${url} — ele parece estar desligado. Inicie com "ollama serve" (ou abra o app Ollama) e tente de novo`,
      )
    }
    throw new Error(`falha ao chamar o Ollama em ${url}: ${err.message}`)
  }

  if (!response.ok) {
    const body = await response.text().catch(() => '')
    if (response.status === 404 && /not found/i.test(body)) {
      throw new Error(
        `modelo "${model}" não está instalado no Ollama local. Instale com: ollama pull ${model}`,
      )
    }
    throw new Error(
      `Ollama respondeu ${response.status} ${response.statusText} em ${url}: ${body.slice(0, 500)}`,
    )
  }

  const payload = await response.json()
  if (typeof payload.response !== 'string') {
    throw new Error(
      `resposta do Ollama não trouxe o campo "response" esperado: ${JSON.stringify(payload).slice(0, 500)}`,
    )
  }
  return payload.response
}

// -------------------------------------------------------------------------
// 3. API — GET /api/insights/numbers + POST /api/insights, os dois
//    autenticados pelo INGEST_TOKEN (Authorization: Bearer). Distingue
//    "não consegui alcançar a API" (fetch lançou — rede/DNS/TLS) de "a API
//    recusou a requisição" (fetch respondeu, response.ok === false).
// -------------------------------------------------------------------------

async function apiRequest(url, { token, fetchImpl, init = {} }) {
  let response
  try {
    response = await fetchImpl(url, {
      ...init,
      headers: {
        ...(init.headers ?? {}),
        authorization: `Bearer ${token}`,
      },
    })
  } catch (err) {
    throw new Error(
      `não consegui alcançar a API em ${url} — confira a conexão e a URL (--api-url ou $FINANCAS_API_URL). Detalhe: ${err.message}`,
    )
  }

  if (!response.ok) {
    const body = await response.text().catch(() => '')
    if (response.status === 401) {
      throw new Error(
        `a API recusou o token (401 em ${url}) — confira se o INGEST_TOKEN deste comando é o MESMO configurado no servidor ("wrangler secret put INGEST_TOKEN" em produção, ou a chave INGEST_TOKEN de .dev.vars em dev)`,
      )
    }
    throw new Error(
      `a API recusou a requisição (${response.status} ${response.statusText} em ${url}): ${body.slice(0, 500)}`,
    )
  }

  return response
}

async function parseEnvelope(response, url) {
  let body
  try {
    body = await response.json()
  } catch (err) {
    throw new Error(
      `resposta da API em ${url} não é JSON válido: ${err.message}`,
    )
  }
  if (
    body === null ||
    typeof body !== 'object' ||
    typeof body.ok !== 'boolean'
  ) {
    throw new Error(
      `resposta da API em ${url} não tem o formato esperado (envelope ok/data/notifications): ${JSON.stringify(body).slice(0, 500)}`,
    )
  }
  if (!body.ok) {
    const mensagem = body.notifications?.[0]?.message ?? 'motivo não informado'
    throw new Error(`a API recusou a requisição em ${url}: ${mensagem}`)
  }
  return body.data
}

export async function fetchNumbers({ apiUrl, token, competence, fetchImpl }) {
  const url = `${apiUrl}/api/insights/numbers?competence=${encodeURIComponent(competence)}`
  const response = await apiRequest(url, { token, fetchImpl })
  return parseEnvelope(response, url)
}

export async function postInsight({
  apiUrl,
  token,
  texto,
  modelo,
  periodo,
  fetchImpl,
}) {
  const url = `${apiUrl}/api/insights`
  const response = await apiRequest(url, {
    token,
    fetchImpl,
    init: {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ texto, modelo, periodo }),
    },
  })
  return parseEnvelope(response, url)
}

// -------------------------------------------------------------------------
// 4. CLI — parsing de argumentos e orquestração
// -------------------------------------------------------------------------

const FLAGS_COM_VALOR = {
  '--competencia': 'competence',
  '--competence': 'competence',
  '--modelo': 'model',
  '--model': 'model',
  '--url': 'ollamaUrl',
  '--api-url': 'apiUrl',
  '--api': 'apiUrl',
}

export function parseArgs(argv) {
  const args = {
    competence: undefined,
    model: undefined,
    ollamaUrl: undefined,
    apiUrl: undefined,
    help: false,
    error: null,
  }

  for (let i = 0; i < argv.length; i++) {
    const atual = argv[i]
    if (atual === '--help' || atual === '-h') {
      args.help = true
      continue
    }
    const campo = FLAGS_COM_VALOR[atual]
    if (campo !== undefined) {
      const valor = argv[i + 1]
      if (valor === undefined) {
        args.error = `a opção ${atual} precisa de um valor`
        return args
      }
      args[campo] = valor
      i++
      continue
    }
    args.error = `opção desconhecida: ${atual}`
    return args
  }

  return args
}

function usage() {
  return `Uso: node scripts/insight.mjs [opções]

Lê os números já calculados de uma competência (GET /api/insights/numbers),
manda pro Ollama local escrever uma leitura em texto por cima deles, e
publica o resultado (POST /api/insights). NÃO envia nenhum lançamento cru
pro modelo — só os totais agregados que a API já calculou.

Opções:
  --competencia, --competence <YYYY-MM>  Competência a resumir (default: mês corrente, fuso de Teresina)
  --modelo, --model <nome>               Modelo do Ollama (default: ${DEFAULT_MODEL})
  --url <url>                            URL do Ollama (default: ${DEFAULT_OLLAMA_URL}, ou $OLLAMA_URL)
  --api-url, --api <url>                 URL base da API (default: ${DEFAULT_API_URL}, ou $FINANCAS_API_URL)
  --help, -h                             Mostra esta ajuda

Requisitos:
  - Ollama rodando localmente (ollama serve) com o modelo instalado (ex.: ollama pull ${DEFAULT_MODEL})
  - Variável de ambiente INGEST_TOKEN definida com o MESMO valor configurado
    no servidor ("wrangler secret put INGEST_TOKEN" em produção, ou a chave
    INGEST_TOKEN de .dev.vars em dev)
`
}

export async function run(argv, deps = {}) {
  const {
    fetchImpl = fetch,
    env = process.env,
    log = console.log,
    logError = console.error,
    now = () => new Date(),
  } = deps

  const args = parseArgs(argv)

  if (args.help) {
    log(usage())
    return 0
  }
  if (args.error) {
    logError(`erro: ${args.error}`)
    log(usage())
    return 2
  }

  const token = env.INGEST_TOKEN
  if (!token) {
    logError(
      'erro: a variável de ambiente INGEST_TOKEN não está definida — este comando não tem como se autenticar na API.\n' +
        '      Defina antes de rodar, com o MESMO valor configurado no servidor via "wrangler secret put INGEST_TOKEN"\n' +
        '      (ou a chave INGEST_TOKEN de apps/financas/.dev.vars em desenvolvimento). Ex.: export INGEST_TOKEN=...',
    )
    return 1
  }

  const competence = args.competence ?? competenciaAtual(now())
  if (!COMPETENCE_RE.test(competence)) {
    logError(`erro: competência inválida (esperado YYYY-MM): ${competence}`)
    log(usage())
    return 2
  }

  const model = args.model ?? DEFAULT_MODEL
  const ollamaUrl = args.ollamaUrl ?? env.OLLAMA_URL ?? DEFAULT_OLLAMA_URL
  const apiUrl = (
    args.apiUrl ??
    env.FINANCAS_API_URL ??
    DEFAULT_API_URL
  ).replace(/\/+$/, '')

  log(`==> buscando os números de ${competence} em ${apiUrl}`)
  let numbers
  try {
    numbers = await fetchNumbers({ apiUrl, token, competence, fetchImpl })
  } catch (err) {
    logError(`erro: ${err.message}`)
    return 1
  }

  const prompt = buildPrompt(numbers)

  log(
    `==> consultando o Ollama (modelo ${model}, ${ollamaUrl}) — pode levar alguns segundos`,
  )
  let rawResponse
  try {
    rawResponse = await callOllama({ model, prompt, url: ollamaUrl, fetchImpl })
  } catch (err) {
    logError(`erro: ${err.message}`)
    return 1
  }

  const texto = rawResponse.trim()
  if (texto === '') {
    logError('erro: o modelo devolveu texto vazio — nada foi publicado.')
    logError('--- saída bruta do modelo ---')
    logError(rawResponse)
    return 1
  }

  log(`==> publicando o insight de ${competence} (modelo ${model})`)
  try {
    await postInsight({
      apiUrl,
      token,
      texto,
      modelo: model,
      periodo: competence,
      fetchImpl,
    })
  } catch (err) {
    logError(`erro: ${err.message}`)
    logError('--- texto gerado (NÃO publicado) ---')
    logError(texto)
    return 1
  }

  log('')
  log(`==> insight de ${competence} publicado com sucesso`)
  log('')
  log(texto)
  return 0
}

if (
  process.argv[1] &&
  import.meta.url === pathToFileURL(process.argv[1]).href
) {
  const codigoDeSaida = await run(process.argv.slice(2))
  process.exitCode = codigoDeSaida
}

File: apps/promeia/src/promeia/insight.py
"""Insight financeiro — porte de apps/financas/scripts/insight.mjs.

⚠️ O MODELO NUNCA VÊ LANÇAMENTO CRU. `GET /api/insights/numbers` devolve só
agregado (totais por categoria/período, já calculados no Worker) — build_prompt
só recebe isso. É o que mantém o escopo do INGEST_TOKEN honesto: lê totais,
escreve prosa, nunca toca o livro-caixa.

⚠️ A ARITMÉTICA NÃO É IA. "Onde gastei mais", "quanto subiu" e "o que mais
cresceu" são consultas exatas, feitas no Worker, e a tela renderiza esses
números direto do dado. O texto gerado aqui é a leitura AO REDOR deles —
nenhum número exibido vem do modelo.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel

from promeia import ollama, ramielle
from promeia.config import ConfigError, Settings
from promeia.dates import competencia_atual, competencia_valida
from promeia.money import format_brl


class InsightVazio(Exception):
    """O modelo devolveu texto vazio — nada foi publicado."""


class PublicacaoFalhou(Exception):
    """Falhou ao publicar DEPOIS de o texto já estar gerado.

    ⚠️ Carrega o texto de propósito. Ele custou uma rodada de modelo, e a
    janela entre gerar e publicar é onde a rede realmente cai. Deixar a
    exceção de transporte subir crua perderia o texto e obrigaria a rodar
    tudo de novo — o CLI Node que este módulo substitui já imprimia o texto
    sob "--- texto gerado (NÃO publicado) ---" por esse motivo.

    NÃO herda de RamielleError: é um conceito do promeia (houve trabalho
    salvável), não uma categoria de falha de transporte. Quem trata precisa
    poder distinguir as duas.
    """

    def __init__(self, texto: str, causa: Exception) -> None:
        super().__init__(str(causa))
        self.texto = texto
        self.causa = causa


# --- prompt -------------------------------------------------------------


def _linha_categoria(row: dict, index: int) -> str:
    return f"{index + 1}. {row['category_name']}: {format_brl(abs(row['total_cents']))}"


def _linha_variacao(numbers: dict) -> str:
    pct = numbers["variation_pct"]
    cents = numbers["variation_cents"]
    anterior = numbers["previous_competence"]
    if pct is None:
        return f"Sem base de comparação: não houve gasto registrado em {anterior}."
    if cents == 0:
        return f"Sem variação em relação a {anterior}."
    direcao = "Aumento" if cents > 0 else "Redução"
    return (
        f"{direcao} de {format_brl(abs(cents))} ({abs(pct)}%) em relação a {anterior}."
    )


def _linha_maior_crescimento(numbers: dict) -> str:
    b = numbers["biggest_increase"]
    if b is None:
        return "(sem dado suficiente para apontar)"
    direcao = "aumento" if b["delta_cents"] >= 0 else "redução"
    return (
        f"{b['category_name']}: foi de {format_brl(abs(b['previous_cents']))} "
        f"para {format_brl(abs(b['current_cents']))} ({direcao} de "
        f"{format_brl(abs(b['delta_cents']))})."
    )


def build_prompt(numbers: dict) -> str:
    """Monta o prompt a partir do payload de GET /api/insights/numbers.

    As "REGRAS OBRIGATÓRIAS" no fim são o que impede o modelo de inventar um
    número. Não as enfraqueça: o prompt é contrato, e o dono lê este texto
    todo mês achando que os números vieram do banco.
    """
    competence = numbers["competence"]
    anterior = numbers["previous_competence"]
    top = numbers["top_categories"]

    linhas_categorias = (
        "\n".join(_linha_categoria(row, i) for i, row in enumerate(top))
        if top
        else "(nenhum gasto registrado nesta competência)"
    )

    return f"""Você escreve um resumo financeiro curto, em português do Brasil, para o dono de um controle financeiro pessoal/PJ.

Abaixo estão os ÚNICOS fatos que você pode usar. Eles já foram calculados por consulta exata ao banco de dados — você NÃO tem acesso a nenhum lançamento individual, só a estes totais já agregados.

Competência: {competence} (mês anterior de comparação: {anterior})
Total gasto em {competence}: {format_brl(abs(numbers["total_cents"]))}
Total gasto em {anterior}: {format_brl(abs(numbers["previous_total_cents"]))}
Variação: {_linha_variacao(numbers)}

Maiores categorias de gasto em {competence} (da maior para a menor):
{linhas_categorias}

Categoria com a maior variação de gasto em relação a {anterior}:
{_linha_maior_crescimento(numbers)}

Escreva um parágrafo de 3 a 5 frases resumindo esses fatos para o dono, em tom direto, sem saudação, sem "espero que ajude", sem markdown, sem lista — só o parágrafo corrido.

REGRAS OBRIGATÓRIAS, sem exceção:
- Use SOMENTE os números e nomes de categoria que aparecem acima. Não calcule nada novo, não arredonde diferente do que já está escrito, não converta nem estime.
- NUNCA invente um valor, uma categoria ou uma porcentagem que não esteja listada acima.
- Se um fato não tem dado suficiente (ex.: "sem base de comparação" ou "nenhum gasto registrado"), diga isso em vez de inventar um número.
- Não é preciso citar todos os números — escolha o que mais importa para contar a história do mês, mas nunca cite um número que não veio da lista acima."""


# --- orquestração -------------------------------------------------------


def run_insight(
    *,
    settings: Settings,
    competence: str | None = None,
    ler: Callable[[str], dict] | None = None,
    gerar: Callable[[str], str] | None = None,
    publicar: Callable[[str, str, str], Any] | None = None,
) -> tuple[str, str]:
    """Lê os números, gera a leitura, publica. Devolve (competência, texto).

    Os três passos são injetáveis: é o que permite testar a ORDEM e os casos
    degenerados sem tocar Ollama nem rede.
    """
    comp = competence or competencia_atual()
    if not competencia_valida(comp):
        raise ValueError(f"competência inválida (esperado YYYY-MM): {comp}")

    # ⚠️ ANTES de qualquer I/O, e antes da rodada de modelo. Com o token vazio
    # nada aqui pode dar certo: `fetch_numbers` monta o header "Bearer " e o
    # httpx o recusa (`Illegal header value b'Bearer '`) — que o `except
    # RequestError` de ramielle.py classificava como RamielleUnreachable,
    # "confira a conexão e a URL", mandando o dono mexer na rede por um
    # problema de CONFIGURAÇÃO. Falhar aqui, com nome próprio, é o conserto da
    # classificação: um erro só é de transporte se uma requisição chegou a ser
    # tentada. Ver Settings.exigir_ingest_token para por que a exigência mora
    # neste ponto e não no boot.
    #
    # `cli.py` já faz a mesma checagem antes de chamar esta função — e continua
    # fazendo, de propósito: a mensagem dele fala a língua de quem está num
    # terminal ("export INGEST_TOKEN=..."), e ele a emite ANTES das linhas
    # "==> buscando os números", que seriam mentira. Esta aqui é a rede para
    # TODO chamador, incluindo o HTTP, que não passa pelo CLI.
    settings.exigir_ingest_token()

    _ler = ler or (
        lambda c: ramielle.fetch_numbers(
            base_url=settings.ramielle_url,
            token=settings.ingest_token,
            competence=c,
        )
    )
    _gerar = gerar or (
        lambda p: ollama.generate(
            model=settings.ollama_model, prompt=p, base_url=settings.ollama_url
        )
    )
    _publicar = publicar or (
        lambda texto, modelo, periodo: ramielle.post_insight(
            base_url=settings.ramielle_url,
            token=settings.ingest_token,
            texto=texto,
            modelo=modelo,
            periodo=periodo,
        )
    )

    numbers = _ler(comp)
    try:
        prompt = build_prompt(numbers)
    except (KeyError, TypeError) as err:
        # `numbers` veio de fora (ramielle.fetch_numbers) — um envelope
        # ok:true sem "data" (numbers vira None: TypeError) ou faltando uma
        # chave esperada (KeyError) não pode virar traceback cru. É recusa
        # do formato da API, não falha de transporte: RamielleRefused, para
        # que a rota e o CLI já saibam tratar sem um `except` novo.
        raise ramielle.RamielleRefused(
            "a API devolveu um payload de números fora do formato esperado "
            f"(GET /api/insights/numbers): {err}"
        ) from err
    texto = _gerar(prompt).strip()

    if texto == "":
        raise InsightVazio(
            "o modelo devolveu texto vazio — nada foi publicado. "
            "Publicar um insight em branco pareceria sucesso e deixaria o card "
            "vazio na tela sem explicação"
        )

    try:
        _publicar(texto, settings.ollama_model, comp)
    except ramielle.RamielleError as err:
        raise PublicacaoFalhou(texto, err) from err

    return comp, texto


# --- rota ---------------------------------------------------------------

router = APIRouter()


class InsightBody(BaseModel):
    competence: str | None = None


def _erro(status: int, code: str, message: str) -> JSONResponse:
    return JSONResponse(
        status_code=status, content={"ok": False, "code": code, "message": message}
    )


@router.post("/insight")
def gerar_insight(body: InsightBody, request: Request):
    """Gera e publica o insight. **Nenhum erro daqui sai como 502 — MEDIDO.**

    ⚠️ A Cloudflare SUBSTITUI o corpo de um 502 da origem pelo próprio error
    page. Medido pelo túnel real (`promeia.piluvitu.com.br`), 3/3 de cada lado,
    e depois isolado por status contra uma origem descartável que só devolvia
    um JSON fixo:

    | Status da origem | Local (:8082)               | Pelo túnel                    |
    | ---------------- | --------------------------- | ----------------------------- |
    | 400 / 422        | application/json            | **JSON intacto**              |
    | **500**          | application/json            | **JSON intacto**              |
    | **502**          | application/json, 162 bytes | text/plain, 16b: `error code: 502` |
    | **503**          | application/json            | **JSON intacto**              |

    Ou seja: **o 502 é o único status cujo corpo o túnel come**, e comer o
    corpo aqui não perde só uma frase bonita — perde o produto. As mensagens
    deste serviço são a feature (`CLAUDE.md` § _Erros são o produto_), e o
    cliente do outro lado (`apps/ramielle/src/lib/promeia.ts#chamarPromeia`)
    trata "erro HTTP sem `code`/`message` no corpo" como **não alcancei o
    Mac** — então todo 502 daqui chegaria ao dono como *"Suba o promeia no
    Mac"*, com o promeia de pé. A pior mensagem possível: manda arrumar o que
    está certo.

    Por isso **todo** erro deste módulo responde 503 (ou 422, para corpo
    inválido), e a regra é grepável: nenhum `502` aparece neste arquivo. A
    distinção que importa — "não alcancei" × "alcancei e recusou" — nunca
    morou no status, mora no campo `code`, que viaja DENTRO do corpo e
    sobrevive junto com ele.

    ⚠️ **`revisao_rotas.py` NÃO foi alinhado a isto, e a razão é contrato, não
    esquecimento**: `/llm/*` é consumido hoje pelo `apps/ramielle`, cujo
    `src/routes/atelier.test.ts` tem um teste explícito
    (`'502 ollama_failed do promeia é repassado como 502, intocado'`) que fixa
    o 502 daquelas rotas. Mudá-lo é uma decisão dos dois lados, com o teste de
    lá junto — não um efeito colateral desta task, que é sobre `/insight`
    (zero consumidor no ramielle: `grep -rn insight apps/ramielle/src` não
    devolve nada).
    """
    settings: Settings = request.app.state.settings
    try:
        competence, texto = run_insight(settings=settings, competence=body.competence)
    except ValueError as err:
        return _erro(422, "invalid_competence", str(err))
    except ConfigError as err:
        # Configuração faltando ≠ falha de transporte. Este ramo existe para o
        # INGEST_TOKEN vazio, que antes saía como `ramielle_unreachable`.
        return _erro(503, "ingest_token_missing", str(err))
    except InsightVazio as err:
        return _erro(503, "empty_insight", str(err))
    except PublicacaoFalhou as err:
        # Devolve o texto no corpo do erro: quem chamou pode republicar sem
        # gastar outra rodada de modelo. Precisa vir ANTES dos ramos de
        # RamielleError — PublicacaoFalhou não herda deles de propósito, mas
        # a ordem deixa a intenção explícita pra quem lê.
        #
        # ⚠️ É o caso que mais dependia de sair do 502: o texto já custou uma
        # rodada de modelo (20-33 s de GPU medidos), viaja em `data.texto`, e
        # com o corpo comido pelo túnel ele sumia — sem ninguém saber que
        # existiu.
        return JSONResponse(
            status_code=503,
            content={
                "ok": False,
                "code": "publish_failed",
                "message": str(err),
                "data": {"texto": err.texto},
            },
        )
    except ollama.OllamaModelMissing as err:
        return _erro(503, "ollama_model_missing", str(err))
    except ollama.OllamaUnreachable as err:
        # O promeia está de pé, o Ollama não. Quem chama precisa dessa
        # distinção pra dizer "abra o Ollama" em vez de "suba o promeia" — e
        # ela viaja no `code`, que sobrevive ao túnel junto com o corpo.
        return _erro(503, "ollama_unreachable", str(err))
    except ollama.OllamaFailed as err:
        return _erro(503, "ollama_failed", str(err))
    except ramielle.RamielleUnreachable as err:
        return _erro(503, "ramielle_unreachable", str(err))
    except ramielle.RamielleRefused as err:
        return _erro(503, "ramielle_refused", str(err))
    # Rede de segurança, DEPOIS de toda classe-folha: uma subclasse nova de
    # OllamaError/RamielleError que ninguém lembrou de listar acima degrada
    # para um erro com mensagem própria em vez de virar 500 cru — a mesma
    # assimetria que cli.py (que já captura as classes-base) não tinha.
    except ollama.OllamaError as err:
        return _erro(503, "ollama_error", str(err))
    except ramielle.RamielleError as err:
        return _erro(503, "ramielle_error", str(err))

    return {
        "ok": True,
        "data": {
            "competence": competence,
            "texto": texto,
            "modelo": settings.ollama_model,
        },
    }
